src/storage-controller/src/healthcheck.rs

```diff
@@ -0,0 +1,467 @@
+        });
+    } else {
+        packer.push(Datum::Null);
+    }
+
+    row
+}
+
+/// Represents an entry in `mz_internal.mz_{source|sink}_status_history`
+/// For brevity's sake, this is represented as a single type because currently
+/// the two relation's have the same shape. If thta changes, we can bifurcate this
+#[derive(Clone)]
+pub struct RawStatusUpdate {
+    /// The ID of the collection
+    pub id: GlobalId,
+    /// The timestamp of the status update
+    pub ts: u64,
+    /// The name of the status update
+    pub status_name: String,
+    /// A succinct error, if any
+    pub error: Option<String>,
+    /// Hints about the error, if any
+    pub hints: BTreeSet<String>,
+    /// The full set of namespaced errors that consolidated into the `error`, if any
+    pub namespaced_errors: BTreeMap<String, String>,
+}
+
+/// A lightweight wrapper around [`CollectionManager`] that assists with
+/// appending status updates to to `mz_internal.mz_{source|status}_history`
+#[derive(Debug, Clone)]
+pub struct CollectionStatusManager<T>
+where
+    T: Timestamp + Lattice + Codec64 + TimestampManipulation,
+{
+    /// Managed collection interface
+    collection_manager: CollectionManager<T>,
+    /// A list of introspection IDs for managed collections
+    introspection_ids: Arc<std::sync::Mutex<BTreeMap<IntrospectionType, GlobalId>>>,
+    previous_statuses: BTreeMap<GlobalId, String>,
+}
+
+impl<T> CollectionStatusManager<T>
+where
+    T: Timestamp + Lattice + Codec64 + TimestampManipulation + From<EpochMillis>,
+{
+    pub fn new(
+        collection_manager: CollectionManager<T>,
+        introspection_ids: Arc<std::sync::Mutex<BTreeMap<IntrospectionType, GlobalId>>>,
+    ) -> Self {
+        Self {
+            collection_manager,
+            introspection_ids,
+            previous_statuses: Default::default(),
+        }
+    }
+
+    fn pack_status_updates(updates: Vec<RawStatusUpdate>) -> Vec<(Row, Diff)> {
+        updates
+            .into_iter()
+            .map(
+                |RawStatusUpdate {
+                     id,
+                     ts,
+                     status_name,
+                     error,
+                     hints,
+                     namespaced_errors,
+                 }| {
+                    let row = pack_status_row(
+                        id,
+                        ts,
+                        &status_name,
+                        error.as_deref(),
+                        &hints,
+                        &namespaced_errors,
+                    );
+
+                    (row, 1)
+                },
+            )
+            .collect()
+    }
+
+    async fn append_updates(&mut self, updates: Vec<RawStatusUpdate>, type_: IntrospectionType) {
+        let source_status_history_id = *self
+            .introspection_ids
+            .lock()
+            .expect("poisoned lock")
+            .get(&type_)
+            .expect("source status history collection to be registered");
```

Comment:
This message should be parameterized by the introspection type, right? It's not necessarily a source